compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1973,7 +1973,21 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] {
                 // instantiated to `Any` belongs to both types.
                 true
               else
-                isSameType(arg1, arg2) // TODO: handle uninstanciated types
+                isSameType(arg1, arg2) || {
+                  // We can only trust a "no" from `isSameType` when both
+                  // `arg1` and `arg2` are fully instantiated.
+                  val fullyInstantiated = new TypeAccumulator[Boolean] {
+                    override def apply(x: Boolean, t: Type) =
+                      x && {
+                        t match {
+                          case _: SkolemType | _: TypeVar | _: TypeParamRef => false
+                          case _ => foldOver(x, t)
+                        }
+                      }
+                  }
+                  !(fullyInstantiated.apply(true, arg1) &&
+                    fullyInstantiated.apply(true, arg2))
```

Comment:
I think this still misses the case where we have an abstract TypeRef that is not a type parameter. Also, it is a bit unsatisfactory that for toplevel TypeRefs and TypeParamRefs we go to their supertypes but here we do not. So if I understand correctly in the case where we have a TypeParamRef A <: Int and a TypeParamRef B <: String, we would diagnose a possible overlap for

    NonVariant[A], NonVariant[B]

but no such overlap for

    CoVariant[A], Covariant[B]

(because for the latter we do a recursive call of `intersecting`).

Reply:
> I think this still misses the case where we have an abstract TypeRef that is not a type parameter.

I changed the test from `tp.symbol.is(TypeParam)` to `tp.symbol.isAbstractOrParamType` and added a test.

>  Also, it is a bit unsatisfactory that for toplevel TypeRefs and TypeParamRefs we go to their supertypes but here we do not. [...]

Indeed, I guess I can also run the covariant test in the invariant case, invariance should give us more ways to prove non overlap, not less!